test.py

```diff
@@ -412,6 +412,16 @@ def __init__(self, test_no, shortname, suite):
         # Unique file name, which is also readable by human, as filename prefix
         self.uname = "{}.{}".format(self.shortname, self.id)
         self.log_filename = os.path.join(suite.options.tmpdir, self.mode, self.uname + ".log")
+        Test._reset(self)
+
+    def reset(self):
+        """Reset this object, including all derived state."""
+        for cls in reversed(self.__class__.__mro__):
```

Comment:
If I understand this code correctly, given an object of a class derived (possibly indirectly) from `Test` - say `BoostTest`- you want to call `_reset()` method on this object for every ancestor class - e.g. `BoostTest`, `UnitTest`, and `Test`.

This is a very... nonstandard way to do this, and I had to stare at this code for a while to understand what it does, and I'm still not sure that I do. I think in such cases, if there's a more "pythonic" way to do it, we should go the pythonic way so it's easier for the reader.

I think the pythonic way (and a more straightforward one) to do that would be to do a similar thing as with `__init__`, that is:
- `Test.reset` implements the reset logic for `Test` (i.e. `self.success = None`)
- `UnitTest.reset` first calls `super().reset(self)`, then does the reset logic for `UnitTest`; in this case it's `pass`, so you don't even need to define a `UnitTest.reset`
- `BoostTest.reset` first calls `super().reset(self)`, then does the reset logic for `BoostTest`

and similarly for other classes.

I don't insist, maybe there's a good reason to do it this way, but please explain.

Reply:
What you describe won't work. Please try it. I spent some time trying it before I arrived to the implementation you see.

Reply:
the core problem is that super().reset(self) will call the derived reset, not super reset, as you expect. In python a base class is linked from a derived class, not from a base object like you may imagine.

E.g. imagine you have an instance of class Derived. That instance would have  link to an instance of its class (Derived), and the instance of class Derived will have a link to class Base. There is no Base object as such.

Reply:
I made a mistake, it's not `super().reset(self)`, it's `super().reset()`.

According to the following experiment, `super().reset()` calls the superclasses' reset function:
```python
class A:
    def f(self):
        print("a")

class B(A):
    def f(self):
        super().f()
        print("b")

class C(B):
    def f(self):
        super().f()
        print("c")

a = A()
b = B()
c = C()

print("a.f()")
a.f()
print("b.f()")
b.f()
print("c.f()")
c.f()

```

It prints:
```
a.f()
a
b.f()
a
b
c.f()
a
b
c
```
So `b.f()` first calls `a.f()` and then `print("b")`.
`c.f()` first calls `b.f()` and then `print("c")`.


I don't understand how that would work:
> the core problem is that super().reset(self) will call the derived reset, not super reset, as you expect

what would "derived" mean in this context?

Reply:
Something like this?
```python
class A:
    def __init__(self):
        A._f(self)

    def f(self):
        A._f(self)

    def _f(self):
        print("a")

class B(A):
    def __init__(self):
        super().__init__()
        B._f(self)

    def f(self):
        super().f()
        B._f(self)

    def _f(self):
        print("b")

class C(B):
    def __init__(self):
        super().__init__()
        C._f(self)

    def f(self):
        super().f()
        C._f(self)

    def _f(self):
        print("c")

print("init a")
a = A()
print("init b")
b = B()
print("init c")
c = C()

print("a.f()")
a.f()
print("b.f()")
b.f()
print("c.f()")
c.f()
```

Reply:
Ok, that works. The drawback is that one re-defines reset() (i.e. f()) in each child... I  can replace the existing code with this if you like.

Reply:
@kbr- just like you I spent some time playing with it before I coded it the way you see it... I don't completely weed out the possibility you'll find a better way but I did validate my solution with a buddy python guru...

Reply:
> You can't really call child reset() from parent init().
But in your original code, you do not call child reset() from parent init. You call the reset specific to this class: `Test._reset(self)`, `UnitTest._reset(self)` and so on. Like in my examples

If you wanted to call child reset, you could:
```python
self.reset()
```
But you use `Test._reset(self)` (not `self._reset()`) to avoid dynamic dispatch.

Reply:
Note that I gave two examples. The first one is if you care about order to be exactly the same as in your original example. The second is shorted but messes up the order (which may or may not be fine).